bukkit/src/main/java/com/github/juliarn/npclib/bukkit/BukkitActionController.java

```diff
@@ -135,14 +145,31 @@ public void handleMove(@NotNull PlayerMoveEvent event) {
     boolean changedWorld = !Objects.equals(from.getWorld(), to.getWorld());
     boolean changedOrientation = from.getYaw() != to.getYaw() || from.getPitch() != to.getPitch();
     boolean changedPosition = from.getX() != to.getX() || from.getY() != to.getY() || from.getZ() != to.getZ();
+    boolean significantMovement = Math.abs(to.getX() - from.getX()) > 0.5 || Math.abs(to.getY() - from.getY()) > 0.5 || Math.abs(to.getZ() - from.getZ()) > 0.5;
+
+    if (!significantMovement) return;
+
+    Player player = event.getPlayer();
+    UUID playerId = player.getUniqueId();
+    long currentTick = System.currentTimeMillis() / 50; // Convert current time to ticks
+
+    // Cooldown check
+    Long lastProcessedTick = playerCooldowns.get(playerId);
+    if (lastProcessedTick != null && currentTick - lastProcessedTick < COOLDOWN_TICKS) {
+      return; // Skip processing if still within the cooldown period
+    }
+    playerCooldowns.put(playerId, currentTick);
 
     // check if any movement happened (event is also called when standing still)
     if (changedPosition || changedOrientation || changedWorld) {
-      Player player = event.getPlayer();
       for (Npc<World, Player, ItemStack, Plugin> npc : this.npcTracker.trackedNpcs()) {
         // check if the player is still in the same world as the npc
         Position pos = npc.position();
-        if (!npc.world().equals(player.getWorld()) || !npc.world().isChunkLoaded(pos.chunkX(), pos.chunkZ())) {
+        World npcWorld = npc.world();
+
+        // Use cached chunk data to check if the chunk is loaded
+        Set<String> loadedChunksInWorld = loadedChunks.get(npcWorld);
+        if (loadedChunksInWorld == null || !loadedChunksInWorld.contains(chunkKey(pos.chunkX(), pos.chunkZ()))) {
           // if the player is tracked by the npc, stop that
           npc.stopTrackingPlayer(player);
           continue;
```

Comment:
(This code isn't visible in this PR, but) there is code below that does: `BukkitPlatformUtil.distance(npc, to)` too often.

This should only be called when `changedPosition` or maybe `changedWorld` is true. If the player changes orientation, their distance to the NPC won't change. This would save on the expensive distance check for players who just turn their heads around.

Reply:
`distance` shouldn't be too expensive as it's just some subtraction and multiplication

Reply:
It's less preferable over distanceSquared for speed, but all the same it's wasteful to do if the only change is orientation, especially since this is extremely hot code.

Reply:
The `distance` method computes a squared distance - it's just not in the name

Reply:
Oh, I misread it as it's your own distance method, not the vanilla `Location.distance()` method. Nevermind then!